Sources/Moya/Task.swift

```diff
@@ -3,39 +3,48 @@ import Foundation
 /// Represents an HTTP task.
 public enum Task {
 
-    /// A request with no additional data.
-    case requestPlain
-
-    /// A requests body set with data.
-    case requestData(Data)
-
-    /// A request body set with `Encodable` type
-    case requestJSONEncodable(Encodable)
-
-    /// A request body set with `Encodable` type and custom encoder
-    case requestCustomJSONEncodable(Encodable, encoder: JSONEncoder)
-
-    /// A requests body set with encoded parameters.
-    case requestParameters(parameters: [String: Any], encoding: ParameterEncoding)
-
-    /// A requests body set with data, combined with url parameters.
-    case requestCompositeData(bodyData: Data, urlParameters: [String: Any])
-
-    /// A requests body set with encoded parameters combined with url parameters.
-    case requestCompositeParameters(bodyParameters: [String: Any], bodyEncoding: ParameterEncoding, urlParameters: [String: Any])
-
-    /// A file upload task.
-    case uploadFile(URL)
-
-    /// A "multipart/form-data" upload task.
-    case uploadMultipart([MultipartFormData])
-
-    /// A "multipart/form-data" upload task  combined with url parameters.
-    case uploadCompositeMultipart([MultipartFormData], urlParameters: [String: Any])
-
-    /// A file download task to a destination.
-    case downloadDestination(DownloadDestination)
-
-    /// A file download task to a destination with extra parameters using the given encoding.
-    case downloadParameters(parameters: [String: Any], encoding: ParameterEncoding, destination: DownloadDestination)
+    /// All different ways to set parameters in an HTTP request's body.
+    public enum BodyParams {
+        /// The given encodable will be set in the request's body without any additional encoding.
+        case raw(Data)
+
+        /// The given encodable will be json encoded in the request's body.
+        case json(Encodable, JSONEncoder = JSONEncoder())
+
+        /// The given encodable will be url encoded in the request's body.
+        case urlEncoded(Encodable, URLEncodedFormEncoder = URLEncodedFormEncoder())
+
+        /// The given encodable will be encoded in the request's body using the provided encoder.
+        ///
+        /// The provided encoder must not be a `URLEncodedFormParameterEncoder` (use `BodyParams.urlEncoded`instead)
+        /// or `JSONParameterEncoder`(use `BodyParams.json` instead). If this is the case, a `MoyaError.encodableMapping` will be raised.
+        case custom(Encodable, ParameterEncoder)
```

Comment:
interesting - so the `ParameterEncoder` is only accepting `Encodable` type as an input, no more parameters as `[String: Any]`? in that case imo this is too big of a change, and the `ParameterEncoding` is not deprecated in 5.0.0, so I would keep using `ParameterEncoding` for now as personally I wouldn't have the energy to create a new `Encodable` type for each endpoint as this might be get tedious pretty fast

Reply:
Dictionaries are automatically encodable if their keys and values are. So ["Hello": "moya"] is Encodable and can be used as parameter, you don't have to create a custom Encodable object.
This is actually was I use in tests.
JSONEncoding needs a valid JSON object, which in fact are all encodables.
URLEncoding uses the String description of values that are not NSNumber, Array of Dictionary.
So I think in a large majority of use cases no changes will be needed, and if so they should be small.

Reply:
ah you're probably correct - in that case nevermind 👍